### PR TITLE
feat(press): a380 pressurization system

### PR DESCRIPTION
**Pull request by mjuhe**
## Summary of Changes

A380 pressurization system. This includes:
- The Cabin Pressure Control System Application, hosted in the CPIOM B
- The Outflow Valve Control Modules. Which have 3 partitions:
    - The Automatic Control Partition (ACP)
    - The Safety and Override Partition (SOP)
    - The Emergency Pressurization Partition (EPP)

## Additional context

Additional tests and potential adjustments will be required when we can do full flights.

Discord username (if different from GitHub): Commander Ga

## Testing instructions

The A320 system hasn't changed so no testing is required.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/mod.rs [hunk 1]
```diff
@@ -3068,37 +2999,63 @@ mod tests {
...
-        #[ignore]
+        fn aft_ofv_close_during_takeoff() {
+            let test_bed = test_bed()
+                .set_on_ground()
+                .iterate(50)
+                .set_takeoff_power()
+                .iterate_with_delta(400, Duration::from_millis(50));
+            assert_eq!(
+                test_bed.aft_outflow_valve_open_amount(),
+                Ratio::new::<percent>(0.)
```
Comment (Gurgel100) [inline fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/mod.rs:3023]:
```suggestion
                Ratio::default()
```

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/mod.rs [hunk 2]
```diff
@@ -3068,37 +2999,63 @@ mod tests {
...
+                .iterate(50)
+                .set_takeoff_power()
+                .iterate_with_delta(400, Duration::from_millis(50))
+                .vertical_speed_of(Velocity::new::<foot_per_minute>(1000.))
+                .then()
+                .command_aircraft_climb(Length::new::<foot>(0.), Length::new::<foot>(10000.))
+                .and()
+                .iterate(10);
+
+            assert!(test_bed.aft_outflow_valve_open_amount() > Ratio::new::<percent>(0.));
```
Comment (Gurgel100) [inline fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/mod.rs:3040]:
```suggestion
            assert!(test_bed.aft_outflow_valve_open_amount() > Ratio::default());
```

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/cpiom_b.rs [hunk 3]
```diff
@@ -609,3 +677,682 @@ impl SimulationElement for VentilationControlSystemApplication {
...
+
+    fn calculate_exterior_vertical_speed(
+        &self,
+        context: &UpdateContext,
+        new_altitude: Length,
+    ) -> Velocity {
+        Velocity::new::<meter_per_second>(
+            (new_altitude.get::<meter>() - self.exterior_flight_altitude.get::<meter>())
+                / context.delta_as_secs_f64(),
+        )
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/cpiom_b.rs:920]: Oh that's nice!

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/cpiom_b.rs [hunk 4]
```diff
@@ -609,3 +677,682 @@ impl SimulationElement for VentilationControlSystemApplication {
...
+        const B: f64 = -1.4089e-13;
+        const C: f64 = -3.3376e-9;
+        const D: f64 = 4.5905e-4;
+
+        let target_aircraft_altitude_foot = (if self.fma_lateral_mode == 20 {
+            self.cruise_altitude
+        } else {
+            self.exterior_flight_altitude
+        })
+        .get::<foot>();
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/cpiom_b.rs:1048]: I like it, thanks for the suggestion.

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/outflow_valve_control_module.rs [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+    epp: EmergencyPressurizationPartition,
+    outflow_valve: OutflowValve,
+    fault: Option<OcsmFault>,
+}
+
+impl OutflowValveControlModule {
+    pub fn new(
+        context: &mut InitContext,
+        outflow_valve_id: usize,
+        powered_by: Vec<ElectricalBusType>,
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/outflow_valve_control_module.rs:68]: Good spot I missed this, the other modules use an array already so this should've been an array from the begining!

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/outflow_valve_control_module.rs [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+                }
+            }
+            false => {
+                if self.stand_by_channel.has_fault() {
+                    Some(OcsmFault::OneChannelFault)
+                } else {
+                    None
+                }
+            }
+        };
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/outflow_valve_control_module.rs:137]: I've changed this to match the other modules (TADD, VCM and FDAC) as they have the same function.

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/outflow_valve_control_module.rs [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+        self.outflow_valve.open_amount()
+    }
+    fn overpressure_relief_valve_inhibit(&self) -> bool {
+        false //Fixme
+    }
+}
+
+impl SimulationElement for OutflowValveControlModule {
+    fn write(&self, writer: &mut SimulatorWriter) {
+        let failure_count = match self.fault {
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/outflow_valve_control_module.rs:172]: Changed here and in the TADD and VCM.

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/mod.rs [hunk 8]
```diff
@@ -803,153 +815,111 @@ impl SimulationElement for A380AirConditioningSystemOverhead {
...
+        // This can be area or ratio and then multiplied by the ofv area
+        let mut sum = Ratio::default();
 
-    fn safety_valve_open_amount(&self) -> Ratio {
-        self.safety_valve.open_amount()
-    }
-}
+        for controller in self.ocsm.iter() {
+            sum += controller.outflow_valve_open_amount();
+        }
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/mod.rs:907]: Ah nice, didn't know you could sum ratios directly.

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/cpiom_b.rs [hunk 9]
```diff
@@ -609,3 +677,682 @@ impl SimulationElement for VentilationControlSystemApplication {
...
+                    0.
+                },
+            ),
+            None => Velocity::new::<foot_per_minute>(0.),
+        }
+    }
+
+    fn calculate_cabin_target_altitude(&self, ocsm: [&impl OcsmShared; 4]) -> Length {
+        if ocsm.iter().any(|o| o.cabin_target_altitude().is_some()) {
+            ocsm[0].cabin_target_altitude().unwrap_or_default()
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/cpiom_b.rs:1018]: In the future I'll have to add a check to see if an OCSM is failed and to get information from more than one. But I'll leave this for a future PR as failures are not implemented yet. 
At the moment this is a working solution.

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/cpiom_b.rs [hunk 10]
```diff
@@ -609,3 +677,682 @@ impl SimulationElement for VentilationControlSystemApplication {
...
+    }
+
+    fn calculate_cruise_vertical_speed(&self) -> Velocity {
+        if (self.cabin_target_altitude - self.cabin_altitude).abs() > Length::new::<foot>(30.) {
+            // This is an to have a smooth altitude capture
+            Velocity::new::<foot_per_minute>(
+                (self.cabin_target_altitude - self.cabin_altitude)
+                    .get::<foot>()
+                    .clamp(C::MAX_DESCENT_RATE, C::MAX_CLIMB_RATE_IN_DESCENT),
+            )
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/cpiom_b.rs:1124]: Yes I'm aware 😄. Just a simplification here to smooth the altitude capturing.

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/ventilation_control_module.rs [hunk 11]
```diff
@@ -247,3 +264,50 @@ impl BulkVentilationControlSystem {
...
+        }
+    }
+
+    fn update(
+        &mut self,
+        ocsm: [&impl OcsmShared; 4],
+        press_overhead: &impl PressurizationOverheadShared,
+    ) {
+        let orvp_open_allowed =
+            !press_overhead.ditching_is_on() && !ocsm[1].overpressure_relief_valve_inhibit();
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/ventilation_control_module.rs:287]: Only OCSM 2 & 3 send information to the Ventilation Control Module. Same reasoning as above for just taking the 2nd one here, but I've added a comment to clarify in the code.

File: fbw-a380x/docs/a380-simvars.md [hunk 12]
```diff
@@ -171,6 +171,33 @@
...
+    - Bool
+    - True when FWC condition for "DIFF PRESS HI" is met (differential pressure between 8.92 and 9.2 PSI)
+
+- A32NX_PRESS_DIFF_PRESS_EXCESSIVE
+    - Bool
+    - True when FWC condition for "EXCESS DIFF PRESS" is met (differential pressure over 9.65 PSI)
+
+- A32NX_PRESS_NEGATIVE_DIFF_PRESS_EXCESSIVE
+    - Bool
+    - True when FWC condition for "EXCESS DIFF PRESS" is met (differential pressure over 9.65 PSI)
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/docs/a380-simvars.md:199]: Good spot! Changed it now to the correct description

**Comment by mjuhe** [comment]
Thanks again @Gurgel100 for the thorough review! Super cool and helpful stuff, I learn more and more with each PR.

All comments addressed now. The unit tests that were already in the code are applicable to the changes, I just removed the `#[ignore]` flag, so now they pretty much mimic the A320 pressurization tests.

File: fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/outflow_valve_control_module.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Gurgel100) [reply fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning/local_controllers/outflow_valve_control_module.rs]: I thought somehow I remember this code :smile:

**Review by Gurgel100 — APPROVED** [review]
LGTM

**Comment by Gurgel100** [comment]
I assume that the FWC constants for checking an exceedance will be moved into the FWC when we have one?

**Comment by mjuhe** [comment]
> I assume that the FWC constants for checking an exceedance will be moved into the FWC when we have one?

No, the CPCS application provides the FWS with the alerts directly, the conditions are not checked within the FWC.

**Pull request merged**